backup: use separate incremental directory by default

benbardin opened the pull request:
backup: use separate incremental directory by default

Previously support was added to optionally cause interactions with a backup to search for its incremental layers in an alternative path.

This change expands that behavior, adding a default path for incremental layers to interactions with backups in a collection. This new default is the '/incrementals' subdirectory of the collection root. This default may be overridden by the prevoiusly-added incremental_storage parameter. This default means that an operator can easily apply different policies to incremental backups retroactively. This behavior also provides an alternative to passing matching incremental_storage parameters to every interaction with those backups, which improves usability.

During the upgrade, when interacting with existing backups, this behavior remains disabled.
After the upgrade, if an existing chain of incremental backups exists in the old default collection, this behavior also remains disabled.
Also after the upgrade, if no such chain exists, or the chain exists in the new default location, the new behavior will be enabled.

Release note (enterprise change): Incremental backups created by `BACKUP INTO` and `BACKUP TO` are now stored by default under the path 'incrementals' within the backup destination, rather than under each backup's path. This enables easier application of cloud-storage provider policies specifically to incremental backups.

Release justification: This feature is very well-tested, and the code refactoring makes it more verifiably correct.

benbardin committed:
backup: use separate incremental directory by default

Previously support was added to optionally cause interactions with a backup to search for its incremental layers in an alternative path.

This change expands that behavior, adding a default path for incremental layers to interactions with backups in a collection. This new default is the '/incrementals' subdirectory of the collection root. This default may be overridden by the prevoiusly-added incremental_storage parameter. This default means that an operator can easily apply different policies to incremental backups retroactively. This behavior also provides an alternative to passing matching incremental_storage parameters to every interaction with those backups, which improves usability.

During the upgrade, when interacting with existing backups, this behavior remains disabled.
After the upgrade, if an existing chain of incremental backups exists in the old default collection, this behavior also remains disabled.
Also after the upgrade, if no such chain exists, or the chain exists in the new default location, the new behavior will be enabled.

Release note (enterprise change): Incremental backups created by BACKUP TO LATEST IN or BACKUP TO are now stored by default under the path 'incrementals' within the backup destination, rather than under each backup's path. This enables easier management of cloud-storage provider policies specifically applied to incremental backups.

Release justification: This feature is very well-tested, and the code refactoring makes it more verifiably correct.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75970)
<!-- Reviewable:end -->

msbutler commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -722,6 +722,7 @@ func resolveBackupManifests(
 	mkStore cloud.ExternalStorageFromURIFactory,
 	from [][]string,
 	incFrom []string,
+	defaultIncFrom []string,
```
given that either len(incFrom) >0 or len(defaultIncFrom)>0, but never both, I wonder if there's a way to refactor this function so only incFrom gets passed in. It seems the logic that makes this hard is the following, and I don't quite understand it:
```
prev, err = FindPriorBackups(ctx, incStores[0], IncludeManifest)
			if len(prev) > 0 {
				incFrom = defaultIncFrom
			} else {
				incFrom = from[0]
				incStores = baseStores
				prev, err = FindPriorBackups(ctx, incStores[0], IncludeManifest)
			}
```
In words, is this saying: first check the default incremental location, and if there aren't any, check the full backup location?

msbutler commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1788,11 +1767,45 @@ func doRestorePlan(
+			return errors.Errorf("RESTORE FROM ... IN can only by used against a single collection path (per-locality)")
+		}
+
+		if defaultIncFrom, err = appendPaths(from[0][:], path.Join(DefaultIncrementalsSubdir, subdir)); err != nil {
```
nit: might be clearer to move this switch statement below len(incfrom), and then add the following comment above the first first `appendPaths` call: append subdirectory to full backup path first, then to candidate incremental storage paths.

again, I wonder if it's possible to only use `incFrom`, instead of both `incFrom` and `defaultIncFrom`

benbardin replied:
I think that `if len` check isn't needed, so I'll remove it - not what you were asking about, but it's just a little bit clearer :)

msbutler replied:
to continue my ramble from my comment above, what if we moved all the logic from `if subdir == "" {} else{}` into it's own function along the lines of `resolveBackupDests(from, subdir, incFrom) (resolvedFrom [][]string, resolvedIncFrom []string, incrementalbackupSource enum )` . The function could even return a backupLocStruct instead.

The one complication here is we'll need to open a store to figure out if we should use the new or old default. But I think this is fine.

We could then use this function in ShowBackups.

Happy to chat through this idea.

msbutler reviewed (commented):
Left a few comments on the backup and restore code. I still need to review the tests and the show backup code. Let's discuss if you have questions!

I think it's also worth writing out the whole policy of destination resolution in the commit message, or maybe by revising the backup destination resolution doc I sent you that Liv originally wrote. I'm still not totally clear what the intended policy is, i.e.:  given the user passes these params for Backup or Restore, this is what happens.

dt commented:
A few nits on commit message style (I promise these become second nature eventually):
- The prefix is not usually enclosed in brackets and slashes in it are generally read as paths within the repo; you'll see commas if you need a separator e.g. `jobs, sql/stats: blah`, but here I think our team typically just uses the common package name "backup:" (or backupccl:" if we're being pedantic) for changes that touch both backup and restore.
- No need for trailing period on title, and ideally is a tad shorter.
- It's usually good, especially if it isn't linking to an issue with more context, to have more contextual info in the commit message about what is changing, why, and what the edge cases (like mixed version) are, since the doc writes read the whole commit message when reviewing the change. We don't, however, need to stuff all that in the release note, which is generally a one-line item in a big bulleted list of what changed since the last release; things like the mixed version handling are generally more granular than the release note covers, but are very good to put in the commit message body above it.

Putting these nits together, I suggest something like:

> backup: use separate incremental directory by default
>
>Previously in <PR> support was added to optionally cause interactions with a backup to search for its incremental layers in an alternative path. This change adds that behavior by default to interactions with backups in a collection, defaulting to the 'incrementals' subdirectory of the collection root, unless overridden by an `incremental_storage` parameter. Doing so by default means that existing backups are already separated by the time an operator realizes they want to apply different policies to these backups, and eliminates the need to pass matching `incremental_storage` parameters to every interaction with those backups, improving the usability compared to the explicit parameter.
> 
> During the upgrade and after the upgrade, when interacting with existing backups, if the have incremental nested directly under them, interactions with and appending to the incremental chain will use that chain. Once the upgrade is finalized, if an existing backup has no existing incremental chain and a new incremental backup is being created, it will start a chain under the new separate prefix.
> 
> Release note (enterprise change): Incremental backups created by BACKUP TO LATEST IN are now stored by default under the path 'incrementals' within the backup collection rather than under each backup's path, allowing easier management of cloud-storage provider policies specifically applied to incremental backups.

benbardin commented:
Thanks David! I updated the message.

benbardin commented on pkg/ccl/backupccl/backup_destination.go:
```diff
@@ -111,80 +112,111 @@ func resolveDest(
+		return "", "", "", nil, nil, err
+	}
+	if !exists {
+		// There's no backup yet, so ddoing a full one now.
```
Done

benbardin commented on pkg/ccl/backupccl/backup_destination.go:
```diff
@@ -111,80 +112,111 @@ func resolveDest(
 
 	// At this point, the defaultURI is the full path for the backup. For BACKUP
-	// INTO, this path includes the chosenSuffix Once this function returns, the
+	// INTO, this path includes the chosenSuffix. Once this function returns, the
```
All for it, but I'd like to do it in a later diff? I actually tried that, but I thought the noise in the diff here made it harder to follow.

benbardin commented on pkg/ccl/backupccl/constants.go:
```diff
@@ -0,0 +1,25 @@
+
+// Describes the location of a set of incremental backups.
+const (
+	None IncrementalsBackupSource = iota
```
Thanks! Done.

benbardin commented on pkg/ccl/backupccl/backup_destination.go:
```diff
@@ -111,80 +112,111 @@ func resolveDest(
+			return "", "", "", nil, nil, err
+		}
+
+		if !incrementalBackupSubdirEnabled || len(priors) != 0 {
```
ok!

benbardin commented on pkg/ccl/backupccl/backup_destination.go:
```diff
@@ -111,80 +112,111 @@ func resolveDest(
+		if !incrementalBackupSubdirEnabled || len(priors) != 0 {
+			incrementalBackupSource = SameAsFull
+		} else {
+			// This feature is disabled while running a mixed-version cluster.
```
That's what `incrementalBackupSubdirEnabled` manages - I moved the comment to the previous stanza, where I think it makes more sense. Look good?

benbardin commented on pkg/ccl/backupccl/constants.go:
```diff
@@ -0,0 +1,25 @@
+
+// The default subdirectory for incremental backups.
+const (
+	DefaultIncrementalsSubdir = "incrementals"
```
Agreed this isn't good, but I have mixed feelings about calling it a `Collection` - it's _not_ a collection, right? Collections are fully resolved, and this is a...subdirectory, which granted has other meanings in this context. Any other names here come to mind?

msbutler commented:
One small nit on the commit message: this feature only applies to backups with the `BACKUP INTO` syntax, not `BACKUP TO LATEST` syntax. (About to do another pass.. couldn't figure out how to leave an inline comment on the commit msg)

msbutler reviewed (commented):
left another round of comments in Backup/Restore code. Again, haven't looked at tests and show backup yet.

msbutler commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1834,10 +1840,24 @@ func doRestorePlan(
+			ctx, &mem, baseStores, mkStore, fullyResolvedBaseDirectory,
+			fullyResolvedIncrementalsDirectory, endTime, encryption, p.User(),
+		)
+	} else {
```
amazing!!

msbutler commented on pkg/ccl/backupccl/incrementals.go:
```diff
@@ -0,0 +1,168 @@
+// make sure to resolve the paths correctly here. We don't want to accidentally
+// correct an unauthorized file path to an authorized one, then write a backup
+// to an unexpected place or print the wrong error message on a restore.
+func JoinURLPath(args ...string) string {
```
i'm not totally clear on the subtle differences between `filepath.Join` and `path.join` and this helper function. Perhaps a few unit tests for this func would help illustrate? Maybe this function belongs in pkg/cloud/uris.go ? Perhaps @dt has more to say on this.

benbardin replied:
Thanks david - good catch with the separator, I didn't think that through.

path.Join doesn't quite work - I tweaked the comments and added some unit tests. This any clearer?

msbutler commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1788,14 +1752,56 @@ func doRestorePlan(
+	//var fullyResolvedIncrementalsDirectory []string
+	var fullyResolvedSubdir string
+
+	if strings.EqualFold(subdir, "LATEST") {
```
nit: sub "LATEST" for latestFileName

msbutler commented on pkg/ccl/backupccl/show.go:
```diff
@@ -248,22 +257,24 @@ func showBackupPlanHook(
-			if err != nil {
-				return err
-			}
 			if strings.EqualFold(subdir, "LATEST") {
```
nit: sub "LATEST" for latestfileName

msbutler reviewed (commented):
huge improvement!!!!! have not reviewed the tests or debug_backup

dt commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -733,203 +685,240 @@ func resolveBackupManifests(
 	}()
 
-	baseManifest, memSize, err := ReadBackupManifestFromStore(ctx, mem, baseStores[0], encryption)
+	_, memSize, err := ReadBackupManifestFromStore(ctx, mem, baseStores[0], encryption)
```
I'm merging #74735 which should let us reduce some of the complexity here around the separate mem and owned/shrink/grow.

benbardin replied:
yeah, I think this load is a mistake. Removed.

benbardin commented on pkg/ccl/backupccl/incrementals.go:
```diff
@@ -0,0 +1,168 @@
+	user security.SQLUsername,
+	execCfg *sql.ExecutorConfig,
+	explicitIncrementalCollections []string,
+	fullBackupCollections []string,
```
You OK if we punt on that for a bit? I would love to clarify our semantics here, but there's a lot of scope in this diff already, I don't _think_ I'm making this worse, and we may want to loop in Kathryn and/or Liv for this.

msbutler replied:
yup, totally fine to punt!

benbardin commented on pkg/ccl/backupccl/backup_destination.go:
```diff
@@ -111,80 +112,111 @@ func resolveDest(
+		return "", "", "", nil, nil, err
+	}
+	if !exists {
+		// There's no backup yet, so ddoing a full one now.
```
fixed

adityamaru commented on pkg/ccl/backupccl/show.go:
```diff
@@ -303,36 +304,45 @@ func showBackupPlanHook(
+			subdir,
+		)
+		incPaths := []string{}
+		var incStore cloud.ExternalStorage
```
incStore was previously being set to `store`. Just want to confirm that sending an unitialized `incStore` below doesn't cause issues.

benbardin replied:
Yes - in the olden days, all the incrementals were in the base directory, so `store` was correct. Now, they may be, or may not, so `incStore` is set with a bit more complexity.

adityamaru reviewed (commented):
flushing out one round of comments

benbardin commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1692,9 +1718,21 @@ func doRestorePlan(
 	}
-	description, err := restoreJobDescription(p, restoreStmt, from, incFrom, restoreStmt.Options,
+	var fromDescription [][]string
+	if len(from) == 1 {
```
Yeah, it's necessary. I've tried to avoid modifying input arguments because I couldn't keep track of what they meant at given points in the code - that means where we _used_ to mutate `from` to include the fully resolved directories, we no longer do. However, behavior is unchanged in the old-style syntax, so passing from there is just fine.

benbardin commented on pkg/ccl/backupccl/show.go:
```diff
@@ -303,36 +304,45 @@ func showBackupPlanHook(
+			[]string{dest},
+			subdir,
+		)
+		incPaths := []string{}
```
yes, good call!

benbardin commented on pkg/ccl/backupccl/backup_destination.go:
```diff
@@ -71,121 +71,116 @@ func resolveDest(
+	// The defaultStore contains a full backup; consequently, we're conducting an incremental backup.
+
+	// Within the chosenSuffix dir, differentiate incremental backups with partName.
+	partName := endTime.GoTime().Format(DateBasedIncFolderName)
```
done!

benbardin commented on pkg/ccl/backupccl/incrementals.go:
```diff
@@ -0,0 +1,197 @@
+		return incPaths, nil
+	}
+	var prev []string
+	var err error
```
done!

benbardin commented:
@stevendanna @msbutler @adityamaru I think I've made all the changes we've talked about, _and_ the build is green! Give this another pass when you're able? Thank you!

msbutler reviewed (commented):
nearly there! I think it's also worth adding a test to show_test to capture the regression Aditya mentioned in slack.

benbardin commented on pkg/ccl/backupccl/backup_destination_test.go:
```diff
@@ -483,16 +499,94 @@ func TestBackupRestoreResolveDestination(t *testing.T) {
+				// dedicated incrementals subdir by default yet.
+				{
+					expectedSuffix := "/2020/12/25-120000.00"
+					expectedIncDir := "/20201225/123000.00"
```
Alas. Fixed.

benbardin commented on pkg/ccl/backupccl/backup_destination_test.go:
```diff
@@ -483,16 +499,94 @@ func TestBackupRestoreResolveDestination(t *testing.T) {
+					writeLatest(t, collectionLoc, expectedSuffix)
 				}
 
+				// A remote incremental into the fourth full backup: BACKUP INTO LATEST
```
Yes, thank you! Fixed the comment.

benbardin commented on pkg/ccl/backupccl/backup_destination_test.go:
```diff
@@ -238,10 +225,10 @@ func TestBackupRestoreResolveDestination(t *testing.T) {
 				{
 					to := localizeURI(t, baseDir, localities)
 					// The full backup should go into the baseDir.
```
Discussed in slack - this doesn't get us much from the UX perspective, but the code is _probably_ cleaner since we don't have to make copies and fork more on syntax type. Your thoughts?

msbutler replied:
just read through the slack thread. I'm ok with this change! To clarify for Kathryn and future users, I'd amend the commit message so it's clear that this change modifies both `BACKUP TO` and `BACKUP INTO`.

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -9813,3 +9813,147 @@ func TestBackupRestoreSystemUsers(t *testing.T) {
+
+		sib := fmt.Sprintf("BACKUP DATABASE fkdb INTO LATEST IN %s WITH incremental_location = %s", dest, inc)
+		sqlDB.Exec(t, sib)
+		sir := fmt.Sprintf("RESTORE DATABASE fkdb FROM LATEST IN %s WITH new_db_name = 'inc_fkdb'", dest)
```
One worry I have is that it's perfectly legal for the user to use the `incremental_location` parameter to get into this undefined state. Given that we can detect this undefined state, what if we throw an error during RESTORE and SHOW BACKUPS if we detect this, and ask the user to pass the `incremental_location` parameter?

benbardin replied:
OK, this I like! Seems like an easy win.

benbardin commented:
@adityamaru @msbutler I think this rounds out the revisions! I fixed the regression Aditya caught (thanks again!), added a unit test, and tweaked a few other things.

I left the commits separate so it's easier to see the differences - if you think we're good here, I'll accept your ✅ , squash and send to David for stability release approval! Thanks!

msbutler reviewed (approved):
LGTM!

dt commented on pkg/ccl/backupccl/show.go:
```diff
@@ -357,36 +358,47 @@ func showBackupPlanHook(
 		}
-		incPaths, err = FindPriorBackups(ctx, incStore, IncludeManifest)
 
+		collection, computedSubdir := CollectionAndSubdir(dest, subdir)
```
When do we need this? If someone does `SHOW BACKUP <collection>/<backup-date>` instead of `SHOW BACKUP <backup-date> IN <collection>` ?

I'm temped to say that with the move to separated per-collection incremental storage, pointing directly to a raw full backup path like the former should be discouraged, or at least, shouldn't try to be magically collection aware, and maybe _just_ shows the full backup at that dest. That implies implying `SHOW BACKUP abc IN col` would show something different (all the incs) than `SHOW BACKUP col/abc` but that seems... maybe okay? i.e. we moved incremtnals to be collection level, and so use the collection syntax to interact with them?

I dunno. Not blocking, just something to chew on / maybe room to remove some code/complexity of supporting both here.

benbardin replied:
Yes, exactly. Aditya caught the regression, discussed in https://cockroachlabs.slack.com/archives/CJXBR3693/p1646276263127099?thread_ts=1646246621.037659&cid=CJXBR3693.

Naively, I don't think I'd expect those two SHOW commands to have different outputs. It seems like a lot of work for a simple `IN` to convey "this is the modern collection syntax," in ways I'd expect to surprise users. But I do agree this code introduces a lot of complexity to get the "it just works" magic, and I would love to think of ways to simplify by reducing functionality.

benbardin commented:
bors r+

kathancox commented:
Hey @benbardin, would you mind updating the release note here, to clarify that this is an update to the `BACKUP INTO` syntax rather than anything to do with the older `BACKUP TO` syntax? (I assume that is the case.) Just for future reference and docs writers working on release notes.

benbardin commented:
> Hey @benbardin, would you mind updating the release note here, to clarify that this is an update to the `BACKUP INTO` syntax rather than anything to do with the older `BACKUP TO` syntax? (I assume that is the case.) Just for future reference and docs writers working on release notes.

It's actually both; I updated the release note to reflect. (Updating the old syntax didn't get us any meaningful functionality, I think, but the code was easier to organize and understand.) Thanks for catching!
